src/compile.c

```diff
@@ -702,13 +726,139 @@ static block bind_matcher(block matcher, block body) {
   return BLOCK(matcher, body);
 }
 
+/* Build a wrapper around a destructuring matcher so that we can chain them
+ * when we have errors.  The approach is as follows:
+ * FORK_OPT NEXT_MATCHER (unless last matcher)
+ * existing_matcher_block
+ * restructure matched variables into a new jv_object
+ * JUMP END
+ *
+ * END will destructure the object returned from a wrapper into all of the
+ * variables contained in the entire destructuring expression and is what the
+ * program's body binds to.
+ */
+static block alternation_wrapper(block submatcher, block *final_matcher, int is_first, int is_last, jv vars_list) {
+    // We need to generate an output object of what succeeded here so we can
+    // bind everyone to those values.
+    block end = gen_subexp(gen_const(jv_object()));
+
+    // Manually restructure the values we extracted into a new object.
```

Comment:
`STOREV` doesn't create a variable though, and `reduce` and `foreach` definitely `STOREV` multiple times.

Ultimately it's `gen_op_var_fresh()` and this code in `src/compile.c:compile()` that creates/allocates a variable:

```
    if ((opcode_describe(curr->op)->flags & OP_HAS_VARIABLE) &&
        curr->bound_by == curr) {
      curr->imm.intval = var_frame_idx++;
      localnames = jv_array_append(localnames, jv_string(curr->symbol));
    }
```

In `gen_reduce()` we see this:

```
        block gen_reduce(block source, block matcher, block init, block body) {
------>block res_var = gen_op_var_fresh(STOREV, "reduce");
------>block update_var = gen_op_bound(STOREV, res_var);
        ...
                                                gen_op_target(FORK, jmp),
                                                jmp,
                                                body,
var update ------------------->update_var)),
        ...
          return BLOCK(gen_op_simple(DUP),
                       init,
var create --->res_var,
                       gen_op_target(FORK, loop),
                       loop,
                       gen_op_bound(LOADVN, res_var));
        }

```

And this is what a reduction looks like after compilation:

```
$ jq -n --debug-dump-disasm 'reduce . as $i (null; $i)'
0000 TOP
0001 DUP
0002 LOADK null
0004 STOREV $reduce:0
0007 FORK 0023
0009 DUPN
0010 STOREV $i:1
0013 LOADVN $reduce:0
0016 LOADV $i:1
0019 STOREV $reduce:0
0022 BACKTRACK
0023 LOADVN $reduce:0
0026 RET

null
$ 
```

An you can see that multiple `STOREV`s to the same variable can work.  The key is that the `STOREV` in the patterns can't be "fresh", otherwise they *will* create new variables and my plan would fail.